### PR TITLE
feat: Sentiment Analysis

### PR DESCRIPTION
**Pull request by tejaswinichile**
Fixes: https://linear.app/chatwoot/issue/CW-2134/introduce-a-gem-informers-to-check-the-sentiment-of-the-text

### CODE REVIEW

File: app/models/concerns/sentiment_analysis_helper.rb
Comment (scmmishra) [inline app/models/concerns/sentiment_analysis_helper.rb:1]: Probably should move this to enterprise as well

File: app/models/concerns/sentiment_analysis_helper.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
+module SentimentAnalysisHelper
+  extend ActiveSupport::Concern
+
+  def opening_sentiments
+    records = incoming_messages.first(average_message_count)
+    average_sentiment(records)
+  end
+
+  def closing_sentiments
```
Comment (scmmishra) [inline app/models/concerns/sentiment_analysis_helper.rb:9]: We should be showing closing sentiments only if the conversation is resolved. Should we check that too

File: app/models/concerns/sentiment_analysis_helper.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  private
+
+  def average_sentiment_label(records)
+    value = records.pluck(:sentiment).sum { |a| a['value'].to_i }
+    value.negative? ? 'negative' : 'positive'
+  end
+
+  def average_sentiment_score(records)
```
Comment (scmmishra) [inline app/models/concerns/sentiment_analysis_helper.rb:28]: This looks good for now, lets see how this works out with real conversations. If it turns out this needs tweaking, maybe we can have a weighted average, say we have 5 conversations, for the opening sentiment the first message will have more weight, and subsequent messages will have lesser and lesser weight

We can use an exponential decay algorithm

```rb
decay_factor = 0.9  # This value should be between 0 and 1
weights = Array.new(n) { |i| (decay_factor ** (n - i - 1)) }
```

For now we can let this implementation stay

File: app/models/concerns/sentiment_analysis_helper.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    total = records.pluck(:sentiment).sum { |a| a['score'].to_f }
+    total / average_message_count
+  end
+
+  def average_message_count
+    incoming_messages.count >= 10 ? 5 : ((incoming_messages.count / 2) - 1)
+  end
```
Comment (scmmishra) [inline app/models/concerns/sentiment_analysis_helper.rb:35]: We can just have a fixed number, say 5 messages, it's fine if the messages overlap. Right now, we need to test this with real-world conversations and then tweak the scores

 @sojan-official @pranavrajs WDYT?

**Review by scmmishra — CHANGES_REQUESTED** [review]

File: app/models/concerns/sentiment_analysis_helper.rb (hunk 3 above)
Comment (pranavrajs) [reply app/models/concerns/sentiment_analysis_helper.rb:35]: Sounds good.

File: app/models/concerns/sentiment_analysis_helper.rb (hunk 1 above)
Comment (tejaswinichile) [reply app/models/concerns/sentiment_analysis_helper.rb:9]: @scmmishra Yes, I have not added the calls from the controller yet or in the `push_event_data` for the conversation. I was thinking, I will add the condition there itself.

File: app/models/concerns/sentiment_analysis_helper.rb (hunk 2 above)
Comment (tejaswinichile) [reply app/models/concerns/sentiment_analysis_helper.rb:28]: > We can use an exponential decay algorithm

Yes, this sounds good. This is for the cloud customer now to check how it turns out. Then we can think about this.

File: app/models/message.rb [hunk 4]
```diff
@@ -365,4 +367,10 @@ def set_conversation_activity
     conversation.update_columns(last_activity_at: created_at)
     # rubocop:enable Rails/SkipsModelValidations
   end
+
+  def update_message_sentiments
+    # override in the enterprise ::Enterprise::SentimentAnalysisJob.perform_later(self)
+  end
```
Comment (scmmishra) [inline app/models/message.rb:373]: @tejaswinichile don't have much context on this, but is it possible to not have this method here at all? CC: @sojan-official

**Review by scmmishra — APPROVED** [review]
One minor comment. Approving the PR anyway, if there's nothing to change feel free to go ahead and merge it.

**Pull request merged**